cmd/kubeadm/app/cmd/phases/controlplanepreparejoin.go

```diff
@@ -0,0 +1,132 @@
+limitations under the License.
+*/
+
+package phases
+
+import (
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	"k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow"
+	cmdutil "k8s.io/kubernetes/cmd/kubeadm/app/cmd/util"
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	certsphase "k8s.io/kubernetes/cmd/kubeadm/app/phases/certs"
+	controlplanephase "k8s.io/kubernetes/cmd/kubeadm/app/phases/controlplane"
+	etcdphase "k8s.io/kubernetes/cmd/kubeadm/app/phases/etcd"
+	kubeconfigphase "k8s.io/kubernetes/cmd/kubeadm/app/phases/kubeconfig"
+	kubeconfigutil "k8s.io/kubernetes/cmd/kubeadm/app/util/kubeconfig"
+)
+
+type controlPlanePrepareData interface {
+	Cfg() *kubeadmapi.InitConfiguration
+}
+
+// NewControlPlanePreparePhase creates a kubeadm workflow phase that implements the preparation of the node to serve a control plane
+func NewControlPlanePreparePhase() workflow.Phase {
+	return workflow.Phase{
+		Name:  "control-plane-prepare",
+		Short: "Prepares the machine for serving a control plane.",
+		Long:  cmdutil.MacroCommandLongDescription,
+		Phases: []workflow.Phase{
+			newCertsSubphase(),
+			newKubeconfigSubphase(),
+			newManifestsSubphase(),
+		},
+	}
+}
+
+func newCertsSubphase() workflow.Phase {
+	return workflow.Phase{
+		Name:  "certs",
+		Short: "Generates the certificates for all the new control plane components",
+		Run:   runCertsPhaseLocal,
+	}
+}
+
+func newKubeconfigSubphase() workflow.Phase {
+	return workflow.Phase{
+		Name:  "kubeconfig",
+		Short: "Generates the kubeconfig for all the new control plane components",
+		Run:   runKubeconfigPhaseLocal,
+	}
+}
+
+func newManifestsSubphase() workflow.Phase {
+	return workflow.Phase{
+		Name:  "manifests",
+		Short: "Generates the manifests for all the new control plane components",
+		Run:   runManifestsPhaseLocal,
+	}
+}
+
+func runCertsPhaseLocal(c workflow.RunData) error {
+	data, ok := c.(controlPlanePrepareData)
+	if !ok {
+		return errors.New("certs phase invoked with an invalid data struct")
+	}
+
+	// Generate missing certificates (if any)
+	return certsphase.CreatePKIAssets(data.Cfg())
+}
+
+func runKubeconfigPhaseLocal(c workflow.RunData) error {
+	data, ok := c.(controlPlanePrepareData)
+	if !ok {
+		return errors.New("kubeconfig phase invoked with an invalid data struct")
+	}
+
+	// Generate kubeconfig files for controller manager, scheduler and for the admin/kubeadm itself
+	// NB. The kubeconfig file for kubelet will be generated by the TLS bootstrap process in
+	// following steps of the join --experimental-control plane workflow
+	if err := kubeconfigphase.CreateJoinControlPlaneKubeConfigFiles(kubeadmconstants.KubernetesDir, data.Cfg()); err != nil {
+		return errors.Wrap(err, "error generating kubeconfig files")
+	}
+
+	return nil
+}
+
+func runManifestsPhaseLocal(c workflow.RunData) error {
+	data, ok := c.(controlPlanePrepareData)
```

Comment:
`nit` we could name it `joinData`, might have more context. (but i'm ok for both)

Reply:
I am keeping `data` as we have been using for the rest of the init subphases.